### PR TITLE
Allow package.json files to be piped into the builder to support typeVersions

### PR DESCRIPTION
**Pull request by rbuckton**
This PR fixes an issue that is currently blocking https://github.com/microsoft/vscode/pull/77835. 

In TypeScript 3.1 we added support for a compiler version compatibility feature known as ["typesVersions"](http://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#version-selection-with-typesversions). When the compiler attempts to resolve an import or type reference directive we now look for a `"typesVersions"` field in the referenced import's `package.json` file. This field indicates path redirections based on the version of the running TypeScript compiler.

Unfortunately, when `gulp-tsb` is run with `noFilesystemLookup` enabled, the compiler is unable to look up the `package.json` file. This is further exacerbated in the case of https://github.com/microsoft/vscode/pull/77835, as vscode includes `node_modules/@types/**/*.d.ts` in the set of files piped to `gulp-tsb`, resulting in colliding declarations due to the inclusion of both `@types/node/index.d.ts` and `@types/node/ts3.2/index.d.ts`.

I've made the following changes to address these issues:

- When acquiring the `rootNames` to be provided to TypeScript, any non .ts/.tsx (or .js/.jsx in the case of `allowJS`) file is removed. This allows `package.json` files to be piped into a `gulp-tsb` builder so that they can be discovered when `noFilesystemLookup` is enabled.
- To support the module resolution needed for `"typesVersions"` when `noFilesystemLookup` is enabled, the `getDirectories` method of `Host` is implemented to emulate a directory list using the file list known to the host.
- I added an `excludeNodeModulesFromRootNames` option to prevent cases such as when both `node_modules/@types/node/index.d.ts` and `node_modules/@types/node/ts3.2/index.d.ts` are included in the `rootNames` (which would effectively ignore the setting for `"typesVersions"` as they are explicitly provided).
- I added a `createWithIConfiguration` function export that allows you to specify the `excludeNodeModulesFromRootNames` option along with a path to a `tsconfig.json` file so as not to clutter the default `create` function export.

### CODE REVIEW

**Comment by rbuckton** [comment]
cc: @jrieken, @weswigham

**Comment by weswigham** [comment]
We should be considering this before the 3.6 TS release so vscode can actually be built, right?

**Comment by rbuckton** [comment]
There are no changes to TypeScript needed to support this. This issue lies solely in `gulp-tsb`.

File: src/utils.ts [hunk 1]
```diff
@@ -195,6 +195,12 @@ export module maps {
         map.remove = multiMapRemove;
         return map;
     }
+    export function createUniqueMultiMap<T>(): MultiMap<T> {
```
Comment (mjbvz) [inline src/utils.ts:198]: Is this function being used?

File: src/builder.ts [hunk 2]
```diff
@@ -748,8 +760,35 @@ namespace WatchApi {
             return process.cwd();
         }
 
-        function getDirectories(path: string): string[] {
-            return !noFileSystemLookup && ts.sys.getDirectories(path);
+        function addTrailingDirectorySeparator(file: string) {
+            return file && file.charAt(file.length - 1) !== '/' ? file + '/' : file;
```
Comment (mjbvz) [inline src/builder.ts:764]: Do we need to use `path.sep` for these checks so that it works properly cross platform?

**Review by mjbvz — APPROVED** [review]

**Comment by jrieken** [comment]
Sorry, but VS Code isn't using gulp-tsb@master but [gulp-tsb@2.0.x](https://github.com/jrieken/gulp-tsb/tree/release/2.0). That means we won't benefit from any of the changes here. 

Can someone explain to me what this effort is all about? It seems that I missed something. Can VS Code not be compiled with TS 3.6 without these or other changes? Is just the `noFilesystemLookup` the issue (which you should be able to change on our side but likely at the cost of performance)?

Generally, we want to get out of gulp-tsb and use the tsc as-is. We only have gulp-tsb because `tsc -w` was never fast enough. (I believe that has changed). Apart from (fast) compile via gulp-tsb we need gulp for copying resources like css or image-files. Tho, that's easy to keep. The theme so far was "no investment as long as gulp-tsb is still runnning" but if gulp-tsb is now falling apart then I would be very much in favour of sunsetting it.

**Comment by rbuckton** [comment]
The approach we've taken in the TypeScript repo is to just call `tsc -b`: https://github.com/microsoft/TypeScript/blob/master/scripts/build/projects.js

Our build just queues up parallel requests to build projects and passes them all at once to a single `tsc -b` invocation so that `tsc -b` can handle building project reference dependencies.

**Comment by jrieken** [comment]
> The approach we've taken in the TypeScript repo is to just call tsc -b:

Given that we only have one project, is the `-b` option just for caching files on disk for faster compile-startup or does the watch-mode also get faster by this?

**Comment by jrieken** [comment]
Follow-up question on https://www.typescriptlang.org/docs/handbook/project-references.html#caveats. Does that apply to syntax errors only or also to semantic errors?

**Comment by weswigham** [comment]
If you don't have multiple projects, `--incremental` is all that's needed now.

**Comment by jrieken** [comment]
Given that `tsc -w` performance doesn't compare to gulp-tsb (yet?) we should probably revisit this PR. 

> Unfortunately, when gulp-tsb is run with noFilesystemLookup enabled, the compiler is unable to look up the package.json 

Let's take a step back. Is enabling file system lookup and disabling piping of `node_modules/@type/...` all that's needed here? That should be doable with little effort.

**Comment by rbuckton** [comment]
> Is enabling file system lookup and disabling piping of node_modules/@type/... all that's needed here? That should be doable with little effort.

Essentially, yes.

**Comment by jrieken** [comment]
Great. I pushed a few changes to gulp-tsb and published it as 4.0.0 - gulp-tsb is now allowed to read files (we see how that affects perf) and things compile with postinstall-deletion-hacks. Closing PR

**Pull request closed without merging**